File: javascript/src/gilded_rose.js
class Item {
  constructor(name, sellIn, quality){
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class Shop {
  constructor(items=[]){
    this.items = items;
  }

  /*
  /!\ Do not change code above this line /!\
  */


  updateQuality() {
    for (var i = 0; i < this.items.length; i++) {
      switch (this.items[i].name) {
        case 'Aged Brie':
          this.updateAgedBrie(this.items[i]);
          break;
        case 'Backstage passes to a TAFKAL80ETC concert':
          this.updateBackstage(this.items[i]);
          break;
        case 'Sulfuras, Hand of Ragnaros':
          break;
        default:
          this.updateUnknown(this.items[i]);
          break;
      }
    }

    return this.items;
  }

  updateUnknown(item) {
    this.decrementSellIn(item);
    if (item.quality > 0) {
      this.decrementQuality(item);
    }
    if (item.sellIn < 0 && item.quality > 0) {
      this.decrementQuality(item);
    }
  }

  updateBackstage(item) {
    if (item.quality < 50) {
      this.incrementQuality(item)
      if (item.sellIn < 11 && item.quality < 50) {
        this.incrementQuality(item)
      }
      if (item.sellIn < 6 && item.quality < 50) {
        this.incrementQuality(item)
      }
    }
    this.decrementSellIn(item);
    if (item.sellIn < 0) {
      this.resetQuality(item);
    }
  }

  updateAgedBrie(item) {
    this.decrementSellIn(item);
    if (item.quality < 50) {
      this.incrementQuality(item);
      if (item.sellIn < 0) {
        this.incrementQuality(item);
      }
    }
  }

  decrementQuality(item) {
    item.quality--;
  }

  incrementQuality(item) {
    item.quality++;
  }

  resetQuality(item) {
    item.quality = 0;
  }

  decrementSellIn(item) {
    item.sellIn--;
  }
}
module.exports = {
  Item,
  Shop
}

File: python/gilded_rose.py
class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name  # item name
        self.sell_in = sell_in  # days left for item to be sold
        self.quality = quality  # number of items left

    def __repr__(self):
        return f"Item(name={self.name} sell_in={self.sell_in}, quality={self.quality})"

    def __str__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)


class GildedRose(object):
    def __init__(self, items):
        self.items = items

    # /!\ Do not change code above this line /!\ #

    def update_quality(self):
        for item in self.items:
            if (
                    item.name != "Aged Brie"
                    and item.name != "Backstage passes to a TAFKAL80ETC concert"
            ):
                if self.checkQualityGreaterThanZero(item):
                    if self.checkIfNotSulphuras(item):
                        self.decreaseItemQuality(item)
            else:
                if item.quality < 50:
                    self.increaseItemQuality(item)
                    if item.name == "Backstage passes to a TAFKAL80ETC concert":
                        self.backstageTickets(item)

            if self.checkIfNotSulphuras(item):
                self.decreaseSellIn(item)
            if item.sell_in < 0:
                self.resetItem(item)
       

    def checkQualityGreaterThanZero(self, item):
        if item.quality > 0:
            return True
        return False


    def increaseItemQuality(self, item):
        item.quality = item.quality+ 1
        # return item.quality

    def decreaseItemQuality(self, item):
        item.quality = item.quality- 1

    def decreaseSellIn(self, item):
        item.sell_in -= 1

    def checkIfNotSulphuras(self, item):
        if item.name != "Sulfuras, Hand of Ragnaros":
            return True
        return False

    def backstageTickets(self, item): # find a better name
        if item.sell_in < 11:
            if item.quality < 50:
                self.increaseItemQuality(item)

        if item.sell_in < 6:
            if item.quality < 50:
                self.increaseItemQuality(item)

    def resetItem(self, item):
        if item.name != "Aged Brie":

            if item.name != "Backstage passes to a TAFKAL80ETC concert":
                if self.checkQualityGreaterThanZero(item):
                    if self.checkIfNotSulphuras(item):
                        self.decreaseItemQuality(item)
            else:
                item.quality = item.quality - item.quality
        else:
            if item.quality < 50:
                self.increaseItemQuality(item)

if __name__ == "__main__":
    main()
